src/mlpack/core/cv/purged_k_fold_cv_impl.hpp

```diff
@@ -0,0 +1,376 @@
+        WeightsType>(k, xs, datasetInfo, ys, numClasses),
+  embargoPercentage(embargoPercentage),
+  intervals(intervals)
+{
+  CheckIntervals();
+}
+
+template<
+    typename MLAlgorithm,
+    typename Metric,
+    typename MatType,
+    typename PredictionsType,
+    typename WeightsType>
+PurgedKFoldCV<
+    MLAlgorithm,
+    Metric,
+    MatType,
+    PredictionsType,
+    WeightsType>::PurgedKFoldCV(
+        size_t k,
+        const MatType& xs,
+        const PredictionsType& ys,
+        const WeightsType& weights,
+        double embargoPercentage,
+        const MatType& intervals) :
+  KFoldCVBase<
+        PurgedKFoldCV<MLAlgorithm, Metric, MatType, PredictionsType, WeightsType>,
+        MLAlgorithm,
+        Metric,
+        MatType,
+        PredictionsType,
+        WeightsType>(k, xs, ys, weights),
+  embargoPercentage(embargoPercentage),
+  intervals(intervals)
+{
+  CheckIntervals();
+}
+
+template<
+    typename MLAlgorithm,
+    typename Metric,
+    typename MatType,
+    typename PredictionsType,
+    typename WeightsType>
+PurgedKFoldCV<
+    MLAlgorithm,
+    Metric,
+    MatType,
+    PredictionsType,
+    WeightsType>::PurgedKFoldCV(
+        size_t k,
+        const MatType& xs,
+        const PredictionsType& ys,
+        const size_t numClasses,
+        const WeightsType& weights,
+        double embargoPercentage,
+        const MatType& intervals) :
+  KFoldCVBase<
+        PurgedKFoldCV<MLAlgorithm, Metric, MatType, PredictionsType, WeightsType>,
+        MLAlgorithm,
+        Metric,
+        MatType,
+        PredictionsType,
+        WeightsType>(k, xs, ys, numClasses, weights),
+  embargoPercentage(embargoPercentage),
+  intervals(intervals)
+{
+  CheckIntervals();
+}
+
+template<
+    typename MLAlgorithm,
+    typename Metric,
+    typename MatType,
+    typename PredictionsType,
+    typename WeightsType>
+PurgedKFoldCV<
+    MLAlgorithm,
+    Metric,
+    MatType,
+    PredictionsType,
+    WeightsType>::PurgedKFoldCV(
+        size_t k,
+        const MatType& xs,
+        const data::DatasetInfo& datasetInfo,
+        const PredictionsType& ys,
+        const size_t numClasses,
+        const WeightsType& weights,
+        double embargoPercentage,
+        const MatType& intervals) :
+  KFoldCVBase<
+        PurgedKFoldCV<MLAlgorithm, Metric, MatType, PredictionsType, WeightsType>,
+        MLAlgorithm,
+        Metric,
+        MatType,
+        PredictionsType,
+        WeightsType>(k, xs, datasetInfo, ys, numClasses, weights),
+  embargoPercentage(embargoPercentage),
+  intervals(intervals)
+{
+  CheckIntervals();
+}
+
+template<
+    typename MLAlgorithm,
+    typename Metric,
+    typename MatType,
+    typename PredictionsType,
+    typename WeightsType>
+size_t PurgedKFoldCV<
+    MLAlgorithm,
+    Metric,
+    MatType,
+    PredictionsType,
+    WeightsType>::ValidationSubsetFirstCol(size_t i) const
+{
+  assert(i < this->k);
+
+  return (this->k - i - 1) * this->binSize;

```

Comment:
So, this is saying that the first column of the validation set is a certain column in `xs`, just the same as `KFoldCV`.  But don't we have to consider the time steps that the sample covers?  Do samples need to be sorted in order of their start time as represented in `intervals`?  I would have imagined that we would instead taken the whole number of time steps in the dataset (e.g. `intervals.max() - intervals.min()`) and split these into `k` folds, and so in this function we would return a time step instead of an index of a column in `xs`.

However, I am not sure which of the possibilities I just wrote is the actual way that this function should be.  I suspect that my confusion would be alleviated if documentation explaining the precise way data is split was added to `cv.md` :smile: